### DISCUSSION TITLE
AutoScriptMigration example

### DISCUSSION THREAD

**Discussion by gkarwchan** [fluentmigrator/fluentmigrator#1413] [Q&A] [answered]
I am looking for a way to embed an SQL script into the project, and the flunetmigrator runner will create (dynamically) a migration for it (without creating a migration class myself).
I see a class called AutoScriptMigration, and wonder what it is?
Does AutoScriptMigration do that?
if yes, is there an example of it?
if no, is what I want achievable?

**Comment by jzabroski** [comment, marked as answer]
I have never used that feature, but the class summary fully describes how you can test it out.

I will say, I think you are conflating dynamically crating a class with dynamic conventions.  AutoScriptMigration still requires you to specify a Migration attribute with a version number.  You can think of AutoScriptMigration as allowing you to write a migration like this (but without the ceremony, as long as you follow rigid embedded resource naming conventions):

```c#
[Migration(0)]
public class M0 : Migration
{
  public void Up()
  {
    Execute.Sql(@"really long script that you would rather have in a SQL file with IntelliSense like RedGate SQL Prompt");
  }
  public void Down()
  {
    Execute.Sql(@"really long script that you would rather have in a SQL file with IntelliSense like RedGate SQL Prompt");
  }
}
```
In this sense, AutoScript Migration is a Template method that lets you convert an Execute.Sql Up/Down script into an Execute.EmbeddedScript but WITHOUT needing to type as much.


If, instead, you want something to run every time, similar to how tools like RoundhousE work, then you would probably need to use Maintenance Migrations.  I would be very careful with using that paradigm.  We used RoundhousE when I worked at erecruit and it caused all kinds of problems.  In fact, I fixed a serious bug in RoundhousE that took six plus months to get released as a hotfix, despite everyone actively contributing to the project realizing how serious the bug was.  I switched to FluentMigrator and never looked back.
